### PR TITLE
Make webhook retry backoff parameters configurable

### PR DESCRIPTION
**Pull request by tkashem**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Currently webhook retry backoff parameters are hard coded, we want to have the ability to configure the backoff parameters for webhook retry logic.  

The objective of this PR is to make the backoff parameters configurable only. So the default that is being used today should still be in effect.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/authorization.go [hunk 1]
```diff
@@ -33,6 +34,15 @@ import (
 	"k8s.io/client-go/tools/clientcmd"
 )
 
+// DefaultAuthorizationWebhookRetryBackoff is the default backoff parameters for
+// authorization webhook used by the apiserver.
+var DefaultAuthorizationWebhookRetryBackoff = wait.Backoff{
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/options/authorization.go:39]: Maybe we can collapse both `DefaultAuthorizationWebhookRetryBackoff` and `DefaultAuthenticationWebhookRetryBackoff` into a shared backoff `DefaultWebhookRetryBackoff`.

File: pkg/kubeapiserver/authenticator/config.go [hunk 2]
```diff
@@ -68,6 +69,8 @@ type Config struct {
 	WebhookTokenAuthnConfigFile string
 	WebhookTokenAuthnVersion    string
 	WebhookTokenAuthnCacheTTL   time.Duration
+	// WebhookRetryBackoff specifies the backoff parameters of webhook retry
```
Comment (deads2k) [inline pkg/kubeapiserver/authenticator/config.go:72]: maybe tell me why I would want to set this.  Our example use-case could be helpful.

File: pkg/kubeapiserver/options/authentication.go [hunk 3]
```diff
@@ -101,9 +102,10 @@ type TokenFileAuthenticationOptions struct {
 
 // WebHookAuthenticationOptions contains web hook authentication options for API Server
 type WebHookAuthenticationOptions struct {
-	ConfigFile string
-	Version    string
-	CacheTTL   time.Duration
+	ConfigFile   string
+	Version      string
+	CacheTTL     time.Duration
+	RetryBackoff wait.Backoff
```
Comment (deads2k) [inline pkg/kubeapiserver/options/authentication.go:108]: maybe tell me why I would want to set this. Our example use-case could be helpful.

File: pkg/kubeapiserver/options/authentication.go [hunk 4]
```diff
@@ -419,6 +425,7 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
 		ret.WebhookTokenAuthnConfigFile = o.WebHook.ConfigFile
 		ret.WebhookTokenAuthnVersion = o.WebHook.Version
 		ret.WebhookTokenAuthnCacheTTL = o.WebHook.CacheTTL
+		ret.WebhookRetryBackoff = o.WebHook.RetryBackoff
```
Comment (deads2k) [inline pkg/kubeapiserver/options/authentication.go:428]: if we do this, why do we need https://github.com/kubernetes/kubernetes/pull/95705/files#diff-97e779171708fff85d7b1ee3fac01336bf351c39066428e12526d0fd7da5c05cR371 ?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go [hunk 5]
```diff
@@ -54,6 +55,9 @@ type DelegatingAuthenticatorConfig struct {
 	APIAudiences authenticator.Audiences
 
 	RequestHeaderConfig *RequestHeaderConfig
+
+	// WebhookRetryBackoff specifies the backoff parameters of webhook retry
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go:59]: Finally found the spot I care about.  maybe tell me why I would want to set this. Our example use-case could be helpful.

Also, I expected to see a `WithCustomRetryBackoff` function like we use for `ResourceBuilderFlags`.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go [hunk 6]
```diff
@@ -54,6 +55,9 @@ type DelegatingAuthenticatorConfig struct {
...
+	WebhookRetryBackoff wait.Backoff
 }
 
 func (c DelegatingAuthenticatorConfig) New() (authenticator.Request, *spec.SecurityDefinitions, error) {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go:63]: I expected a default backoff wired here.

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 7]
```diff
@@ -102,49 +112,43 @@ func newGenericWebhook(scheme *runtime.Scheme, codecFactory serializer.CodecFact
...
-		if err != nil {
-			return false, err
+		if webhookErr != nil {
+			return false, webhookErr
 		}
 		return true, nil
 	})
-	return err
+
+	if webhookErr == nil && err == wait.ErrWaitTimeout {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go:150]: pull this change into a separate PR.  it looks good, but it's distinct from the overall change.

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 8]
```diff
@@ -47,34 +45,34 @@ type tokenReviewer interface {
...
 }
 
 // New creates a new WebhookTokenAuthenticator from the provided kubeconfig
 // file. It is recommend to wrap this authenticator with the token cache
 // authenticator implemented in
 // k8s.io/apiserver/pkg/authentication/token/cache.
-func New(kubeConfigFile string, version string, implicitAuds authenticator.Audiences, customDial utilnet.DialFunc) (*WebhookTokenAuthenticator, error) {
-	tokenReview, err := tokenReviewInterfaceFromKubeconfig(kubeConfigFile, version, customDial)
+func New(kubeConfigFile string, version string, implicitAuds authenticator.Audiences, retryBackoff wait.Backoff, customDial utilnet.DialFunc) (*WebhookTokenAuthenticator, error) {
+	tokenReview, err := tokenReviewInterfaceFromKubeconfig(kubeConfigFile, version, webhook.DefaultRetryBackoff(0), customDial)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go:66]: why isn't the passed backoff used here?

Comment (deads2k) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go:66]: zero looks like a bug here.

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go [hunk 9]
```diff
@@ -604,3 +604,32 @@ func poller(interval, timeout time.Duration) WaitFunc {
 		return ch
 	})
 }
+
+// ContextBoundExponentialBackoff works with a request context and a Backoff. It ensures that the retry wait never
+// exceeds the deadline specified by the request context.
+func ContextBoundExponentialBackoff(ctx context.Context, backoff Backoff, condition ConditionFunc) error {
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:610]: `ExponentialBackoffWithContext` is more in keeping with the rest of the names.

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go [hunk 10]
```diff
@@ -604,3 +604,32 @@ func poller(interval, timeout time.Duration) WaitFunc {
...
+	for backoff.Steps > 0 {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		if ok, err := condition(); err != nil || ok {
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:618]: `runConditionWithCrashProtection` I would think

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go (hunk 9 above)
Comment (seh) [reply staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:610]: This would have helped with [this recent search of mine](https://kubernetes.slack.com/archives/C0EG7JC6T/p1601514495044600) (from the "sig-api-machinery" Slack channel).

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go (hunk 5 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go]: @deads2k I added the `WithCustomRetryBackoff` function in `DelegatingAuthenticationOptions` - https://github.com/kubernetes/kubernetes/pull/95705/commits/5df6f2d43697b3076438f46400b3bdfeff0b8d84#diff-249fc1fe96479271dbe3cbd4847642b338cfc33a10f1f033fa606da704462320R213-R217. does that sound correct?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go [hunk 11]
```diff
@@ -79,7 +85,13 @@ func (c DelegatingAuthenticatorConfig) New() (authenticator.Request, *spec.Secur
 	}
 
 	if c.TokenAccessReviewClient != nil {
-		tokenAuth, err := webhooktoken.NewFromInterface(c.TokenAccessReviewClient, c.APIAudiences)
+		// Provide a default if WebhookRetryBackoff has not been set by the user.
+		retryBackoff := c.WebhookRetryBackoff
+		if retryBackoff == nil {
+			retryBackoff = webhooktoken.DefaultRetryBackoff()
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go:91]: Is this being left empty somewhere?  I would expect the options to already have defaults set so this is unnecessary.

File: staging/src/k8s.io/apiserver/pkg/authorization/authorizerfactory/delegating.go [hunk 12]
```diff
@@ -35,12 +36,24 @@ type DelegatingAuthorizerConfig struct {
...
+
+	// WebhookRetryBackoff specifies the backoff parameters for the authorization webhook retry logic.
+	// This allows us to configure the sleep time at each iteration and the maximum number of retries allowed
+	// before we fail the webhook call in order to limit the fan out that ensues when the system is degraded.
+	WebhookRetryBackoff *wait.Backoff
 }
 
 func (c DelegatingAuthorizerConfig) New() (authorizer.Authorizer, error) {
+	// Provide a default if WebhookRetryBackoff has not been set by the user.
+	retryBackoff := c.WebhookRetryBackoff
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/authorization/authorizerfactory/delegating.go:48]: I'd rather fail on this and return a clear error.  This is a coding error and most people should be using our recommended options which should have this value set.

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go [hunk 13]
```diff
@@ -36,9 +37,9 @@ const (
 	// PluginName is the name of this plugin, to be used in help and logs.
 	PluginName = "webhook"
 
-	// DefaultInitialBackoff is the default amount of time to wait before
+	// DefaultInitialBackoffDelay is the default amount of time to wait before
 	// retrying sending audit events through a webhook.
-	DefaultInitialBackoff = 10 * time.Second
+	DefaultInitialBackoffDelay = 10 * time.Second
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go:42]: wow that's long.  Leave it for now, but wow.

**Comment by deads2k** [comment]
Please squash what you have and then add the nil check fixes that should be errors in a fresh commit.  This is close.

File: staging/src/k8s.io/apiserver/pkg/authorization/authorizerfactory/delegating.go (hunk 12 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/authorization/authorizerfactory/delegating.go]: @deads2k I agree, I fixed it, we throw clear error now. I originally misinterpreted your comment https://github.com/kubernetes/kubernetes/pull/95705#discussion_r511133119 to add defaulting here.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go (hunk 11 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go]: @deads2k yes, somewhere it's left empty. If I don't default here, all the `kind` tests in ci fail. We are throwing error now. So I am hoping I can find the spot.

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 14]
```diff
@@ -36,6 +37,17 @@ import (
 	openapicommon "k8s.io/kube-openapi/pkg/common"
 )
 
+// DefaultAuthWebhookRetryBackoff is the default backoff parameters for
+// both authentication and authorization webhook used by the apiserver.
```
Comment (sftim) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:41]: If a webhook returns a 5xx response that has `Retry-After` set, the client end (Kubernetes) does not pay attention to that.

Would it be OK to briefly mention in a comment that this implementation does not adjust behavior based on `Retry-After`?
(to be clear - it's OK not to check that header; maybe a future PR can start checking that response header)
